src/write.rs

```diff
@@ -21,115 +21,191 @@
-                CrcWriter::new(DeflateEncoder::new(&mut compressed, Compression::best()));
-            encoder.write(&self.buffer)?;
-            encoder.crc().sum()
-        };
-        let compressed_len = compressed.len() + 19 + 6;
-
-        let buflen = self.buffer.len();
-        let header = [
-            0x1f,
-            0x8b,
-            0x08,
-            0x04,
-            0x00,
-            0x00,
-            0x00,
-            0x00,
-            0x02,
-            0xff,
-            0x06,
-            0x00,
-            66,
-            67,
-            0x02,
-            0x00,
-            (compressed_len & 0xff) as u8,
-            ((compressed_len >> 8) & buflen) as u8,
-        ];
-        let wrote_bytes = self.writer.write(&header[..])?;
-        if wrote_bytes != header.len() {
-            return Err(io::Error::new(io::ErrorKind::Other, "Cannot write header"));
-        }
-
-        let wrote_bytes = self.writer.write(&compressed[..])?;
-        if wrote_bytes != compressed.len() {
-            return Err(io::Error::new(
-                io::ErrorKind::Other,
-                "Cannot write compressed data",
-            ));
-        }
-
-        let crc_bytes = super::bytes_le_u32(crc);
-        let wrote_bytes = self.writer.write(&crc_bytes[..])?;
-        if wrote_bytes != crc_bytes.len() {
-            return Err(io::Error::new(
-                io::ErrorKind::Other,
-                "Cannot write CRC data",
-            ));
-        }
-
-        let buflen_bytes = super::bytes_le_u32(self.buffer.len() as u32);
-        let wrote_bytes = self.writer.write(&buflen_bytes[..])?;
-        if wrote_bytes != buflen_bytes.len() {
-            return Err(io::Error::new(
-                io::ErrorKind::Other,
-                "Cannot write CRC data",
-            ));
-        }
-
-        self.buffer.clear();
+        let (r_buf, s_buf, s_out) = self.channels.as_mut().unwrap();
+        let mut buffer: Vec<u8> = r_buf.recv().unwrap()?;
+        let mut compressed = r_buf.recv().unwrap()?;
+        let s_buf = s_buf.clone();
+        let s_out = s_out.clone();
+        std::mem::swap(&mut self.buffer, &mut buffer);
+        let seq = self.seq;
+        self.pool.spawn(move || {
+            let func = || {
+                let crc = {
+                    let mut encoder =
+                        CrcWriter::new(DeflateEncoder::new(&mut compressed, Compression::best()));
+                    encoder.write(&buffer)?;
+                    encoder.crc().sum()
+                };
+                let buflen = buffer.len();
+                buffer.clear();
+                s_buf.send(Ok(buffer)).unwrap();
+                Ok((buflen, compressed, crc))
+            };
+            s_out.send((seq, func())).unwrap();
+        });
+        self.seq += 1;
         Ok(())
     }
 }
 
-impl<R: io::Write> Drop for BGzWriter<R> {
+impl Drop for BGzWriter {
     fn drop(&mut self) {
         self.flush().unwrap();
```

Comment:
this will panic if there's an IO error.  you need to call flush yourself to guarantee that you get the error.  Not sure what the recommended practice is here.

Reply:
The flush now only reports whether the buffer channel could be received upon. Frankly, if it can't, we should probably be panicking. Perhaps I should change that `?` into an unwrap to fail immediately?

Reply:
Ignore the above, I'm wrong.